common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp

```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                } else {
+                    wpa_psk.assign(bss_it->network_key);
+                }
+                ieee80211w.assign("0");
+                disable_pmksa_caching.assign("1");
+                okc.assign("0");
+                wpa_disable_eapol_key_retries.assign("0");
+            } else {
+                LOG(ERROR) << "Autoconfiguration: " << vap << " invalid authentication type";
+                abort = true;
+                return;
+            }
+
+            LOG(DEBUG) << "Autoconfiguration for ssid: " << bss_it->ssid
+                       << " auth_type: " << auth_type << " encr_type: " << enc_type
+                       << " network_key: " << bss_it->network_key
+                       << " fronthaul=" << beerocks::string_utils::bool_str(bss_it->fronthaul)
+                       << " backhaul=" << beerocks::string_utils::bool_str(bss_it->backhaul);
+
+            conf.set_create_vap_value(vap, "wps_state", bss_it->fronthaul ? "2" : "");
+            conf.set_create_vap_value(vap, "wps_independent", "0");
+            conf.set_create_vap_value(vap, "sFourAddrMode", bss_it->backhaul ? "1" : "");
+            conf.set_create_vap_value(vap, "max_num_sta", bss_it->backhaul ? "1" : "");
+
+            // oddly enough, multi_ap_backhaul_wpa_passphrase has to be
+            // quoted, while wpa_passphrase does not...
+            if (bss_it->fronthaul && !backhaul_wps_ssid.empty()) {
+                conf.set_create_vap_value(vap, "multi_ap_backhaul_ssid",
+                                          "\"" + backhaul_wps_ssid + "\"");
+                conf.set_create_vap_value(vap, "multi_ap_backhaul_wpa_passphrase",
+                                          backhaul_wps_passphrase);
+            }
+
+            conf.set_create_vap_value(vap, "wpa", wpa);
+            conf.set_create_vap_value(vap, "okc", okc);
+            conf.set_create_vap_value(vap, "wpa_key_mgmt", wpa_key_mgmt);
+            conf.set_create_vap_value(vap, "wpa_pairwise", wpa_pairwise);
+            conf.set_create_vap_value(vap, "wpa_psk", wpa_psk);
+            conf.set_create_vap_value(vap, "ieee80211w", ieee80211w);
+            conf.set_create_vap_value(vap, "wpa_passphrase", wpa_passphrase);
+            conf.set_create_vap_value(vap, "disable_pmksa_caching", disable_pmksa_caching);
+            conf.set_create_vap_value(vap, "wpa_disable_eapol_key_retries",
+                                      wpa_disable_eapol_key_retries);
+
+            // finally enable the vap (remove any previously set start_disabled)
+            conf.set_create_vap_value(vap, "start_disabled", "");
+
+        } else {
+            // no more data in the input bss-conf list
+            // disable the rest of the vaps
+            conf.set_create_vap_value(vap, "start_disabled", 1);
+            conf.set_create_vap_value(vap, "ssid", "");
+        }
+    };
+
+    conf.for_all_ap_vaps(configure_func, bss_info_conf_list.begin(), bss_info_conf_list.end(),
```

Comment:
I think it could be very handy if the *abort* mechanism was generic. This can be easily achieved if `configure_func` returns a bool and `for_all_ap_vaps` stops iterating if the function returns false. 
`for_all_ap_vaps` can return true if it has iterated until the end or false if aborted.

Reply:
I am so glad you noticed it. In general I agree with you, but because of implementation decisions I made, I left it like this. 
Internally I use [std::for_each](https://en.cppreference.com/w/cpp/algorithm/for_each), that does not support any kind of [stop_token](https://en.cppreference.com/w/cpp/thread/stop_token) or similar.
So I was left with two options: 
* implement my own iteration and stop on error and return an error.
* use the abort mechanism.

I preferred the later because the internal implementation is simpler this way and the interface is consistence with `std`.